instrumentation/servlet/servlet-3.0/src/main/java/io/opentelemetry/auto/instrumentation/servlet/v3_0/AsyncContextInstrumentation.java

```diff
@@ -105,13 +107,16 @@ public static boolean enter(
         request.setAttribute(CONTEXT_ATTRIBUTE, currentContext);
       }
 
-      return true;
+      // AsyncContext.dispatch() is known to sometimes leak context into Tomcat's thread pool,
+      // and we don't need context propagation downstream from here anyways
+      return withScopedContext(Context.ROOT);
```

Comment:
I found this in the javadoc

This method returns immediately after passing the request and response objects to a container managed thread, on which the dispatch operation will be performed.

Just curious do you know where the context is restored for the business logic on the tomcat thread? Would it just be the servlet handler?

Also it seems like this is a pattern for preventing the automatic copying of context to another thread by concurrent instrumentation. Perhaps we can have a more explicit pattern though.

Reply:
> Just curious do you know where the context is restored for the business logic on the tomcat thread? Would it just be the servlet handler?

See up a few lines, the current context is stored into a ServletRequest attribute, and then later it is restored in `Servlet3Advice`.

Reply:
Ah thanks. IIUC, it's functionalliy the same to move this into that if and return `null` here - that might be more readable.

Reply:
it seems worth preventing other things in the context from leaking here in case there's no span but there's other stuff (not that this is common though)